affinity/cpp-20/d0796r3.md

```diff
@@ -311,34 +329,44 @@ int main(int argc, char * argv[]) {
+
+An allocator capable of allocating memory in the memory region of the **memory resource** represented by a `memory_resource` object by constructing a `pmr::polymorphic_allocator` from the `memory_resource` object.
 
 ### Querying relative affinity
 
-The `affinity_query` class template provides an abstraction for a relative affinity value between two `execution_resource`s. This value depends on a particular `affinity_operation` and `affinity_metric`. As a result, the `affinity_query` is templated on `affinity_operation` and `affinity_metric`, and is constructed from two `execution_resource`s. An `affinity_query` is not meant to be meaningful on its own. Instead, users are meant to compare two queries with comparison operators, in order to get a relative magnitude of affinity. If necessary, the value of an `affinity_query` can also be queried through `native_affinity`, though the return value of this is implementation defined.
+The `affinity_query` class template provides an abstraction of the relative affinity between an `execution_resource` and a `memory_resource` for a particular `affinity_operation` and `affinity_metric`. The `affinity_query` takes the `affinity_operation` and `affinity_metric` as template parameters, and is constructed from an `execution_resource` and a `memory_resource`.
+
+An `affinity_query` is not generally not meaningful on its own. Instead, users are meant to compare two `affinity_query`s via comparison operators, in order to get a relative magnitude of affinity. If necessary, the underlying value of an `affinity_query` can be queried through `native_affinity`, though the return value of this is implementation defined.
 
-Below *(listing 4)* is an example of how to query the relative affinity between two `execution_resource`s.
+Below *(listing 4)* is an example of how to query the relative affinity between an `execution_resource` and a `memory_resource`.
 
 ```cpp
 auto systemResource = this_system::discover_topology();
 
-auto relativeLatency01 = execution::affinity_query<execution::affinity_operation::read,
-  execution::affinity_metric::latency>(systemResource[0], systemResource[1]);
+auto relativeLatency0 = execution::affinity_query<execution::affinity_operation::read,
+  execution::affinity_metric::latency>(systemResource[0], systemResource.memory_resource());
 
-auto relativeLatency02 = execution::affinity_query<execution::affinity_operation::read,
-  execution::affinity_metric::latency>(systemResource[0], systemResource[2]);
+auto relativeLatency1 = execution::affinity_query<execution::affinity_operation::read,
+  execution::affinity_metric::latency>(systemResource[1], systemResource.memory_resource());
 
-auto relativeLatency = relativeLatency01 > relativeLatency02;
+auto relativeLatency = relativeLatency1 > relativeLatency0;
```

Comment:
What are the requirements for the returned type of the comparison operators? Must they be bool?

Reply:
Classes like `less<T>` break if the comparison operators don't return `bool`, because their `operator()` expects to return `bool`.  This issue came up with bitmasks in the SIMD proposal; the author avoided the issue by not overloading comparison operators to return non-`bool`, and instead introducing two-letter comparison operators that return a bitmask for use e.g., in `where` clauses.

Reply:
See e.g., *[comparisons.less]*.  I also read *[expr.rel]* as requiring that `operator<` always return `bool`.

Reply:
The return type of the comparison operators is currently a `std::expected<size_t, error_type>`, the reason for this was so that we could reflect a magnitude rather than simply a boolean as the relative affinity, and also that we want to handle failures to retrieve affinity. However, there have been some concerns about this interface not playing well with others, and this is another good example of that. I have an issue open for looking into alternative solutions to this - https://github.com/codeplaysoftware/standards-proposals/issues/68.